### PR TITLE
plugins/tmux-navigator: port to mkVimPlugin & update available options

### PR DESCRIPTION
**Pull request by MattSturgeon**
This PR ports `tmux-navigator` to the new `mkVimPlugin` system and adds missing options.

The option documentation was re-written based on the docs in [upstream's README](https://github.com/christoomey/vim-tmux-navigator?tab=readme-ov-file#configuration).

I have some questions on the best way to approach some stuff:

- Is there a way to add additional documentation to the actual plugin module?
  - _Done, using #1297_
- What type should `save_on_switch` use?
  - Is there a way to map better named enum values to `1` & `2`?
  - Is this even desired?
- Should I add myself to maintainers?
  - _Done_

### CODE REVIEW

File: plugins/utils/tmux-navigator.nix [hunk 1]
```diff
@@ -6,37 +6,126 @@
...
+            options.silent = true;
+          }
+          {
+            key = "<C-w>l";
+            action = "<cmd><C-U>TmuxNavigateRight<CR>";
+            options.silent = true;
+          }
+          {
+            key = "<C-w>\\";
+            action = "<cmd><C-U>TmuxNavigatePrevious<CR>";
```
Comment (MattSturgeon) [inline plugins/utils/tmux-navigator.nix:122]:
```suggestion
            action = "<CMD><C-U>TmuxNavigatePrevious<CR>";
```

File: plugins/utils/tmux-navigator.nix [hunk 2]
```diff
@@ -6,37 +6,126 @@
...
+    globalPrefix = "tmux_navigator_";
+
+    maintainers = [maintainers.GaetanLepage];
+
+    imports = deprecations;
+
+    settingsOptions = {
+      # FIXME can this be enum of ["buffer" "all"]?
+      save_on_switch = helpers.mkNullOrOption (types.enum [1 2]) ''
+        You can configure the plugin to write the current buffer, or all buffers, when navigating from Vim to tmux.
```
Comment (MattSturgeon) [inline plugins/utils/tmux-navigator.nix:47]: `vim`, `Vim`, `tmux` capitalization is a little inconsistent.

Comment (GaetanLepage) [reply plugins/utils/tmux-navigator.nix:47]: I would go for "vim" here, but this is not too important.

File: plugins/utils/tmux-navigator.nix [hunk 3]
```diff
@@ -6,37 +6,126 @@
...
-  };
-
-  config = mkIf cfg.enable {
-    extraPlugins = [cfg.package];
-
-    globals = {
-      tmux_navigator_save_on_switch = cfg.tmuxNavigatorSaveOnSwitch;
-      tmux_navigator_disable_when_zoomed = cfg.tmuxNavigatorDisableWhenZoomed;
-      tmux_navigator_no_wrap = cfg.tmuxNavigatorNoWrap;
+  # Added 2024-03-19
```
Comment (GaetanLepage) [inline plugins/utils/tmux-navigator.nix:9]:
```suggestion
  # TODO: Introduced 2024-03-19, remove on 2024-05-19
```

File: plugins/utils/tmux-navigator.nix [hunk 4]
```diff
@@ -6,37 +6,126 @@
...
+        bind-key -n 'C-h' if-shell "$is_vim" { send-keys C-h } { if-shell -F '#{pane_at_left}'   {} { select-pane -L } }
+        bind-key -n 'C-j' if-shell "$is_vim" { send-keys C-j } { if-shell -F '#{pane_at_bottom}' {} { select-pane -D } }
+        bind-key -n 'C-k' if-shell "$is_vim" { send-keys C-k } { if-shell -F '#{pane_at_top}'    {} { select-pane -U } }
+        bind-key -n 'C-l' if-shell "$is_vim" { send-keys C-l } { if-shell -F '#{pane_at_right}'  {} { select-pane -R } }
+
+        bind-key -T copy-mode-vi 'C-h' if-shell -F '#{pane_at_left}'   {} { select-pane -L }
+        bind-key -T copy-mode-vi 'C-j' if-shell -F '#{pane_at_bottom}' {} { select-pane -D }
+        bind-key -T copy-mode-vi 'C-k' if-shell -F '#{pane_at_top}'    {} { select-pane -U }
+        bind-key -T copy-mode-vi 'C-l' if-shell -F '#{pane_at_right}'  {} { select-pane -R }
+        ```
```
Comment (GaetanLepage) [inline plugins/utils/tmux-navigator.nix:90]:
```suggestion
        ```shell
          is_vim="ps -o state= -o comm= -t '#{pane_tty}' \
            | grep -iqE '^[^TXZ ]+ +(\\S+\\/)?g?(view|l?n?vim?x?|fzf)(diff)?$'"
  
          bind-key -n 'C-h' if-shell "$is_vim" { send-keys C-h } { if-shell -F '#{pane_at_left}'   {} { select-pane -L } }
          bind-key -n 'C-j' if-shell "$is_vim" { send-keys C-j } { if-shell -F '#{pane_at_bottom}' {} { select-pane -D } }
          bind-key -n 'C-k' if-shell "$is_vim" { send-keys C-k } { if-shell -F '#{pane_at_top}'    {} { select-pane -U } }
          bind-key -n 'C-l' if-shell "$is_vim" { send-keys C-l } { if-shell -F '#{pane_at_right}'  {} { select-pane -R } }
  
          bind-key -T copy-mode-vi 'C-h' if-shell -F '#{pane_at_left}'   {} { select-pane -L }
          bind-key -T copy-mode-vi 'C-j' if-shell -F '#{pane_at_bottom}' {} { select-pane -D }
          bind-key -T copy-mode-vi 'C-k' if-shell -F '#{pane_at_top}'    {} { select-pane -U }
          bind-key -T copy-mode-vi 'C-l' if-shell -F '#{pane_at_right}'  {} { select-pane -R }
        ```
```
(indentation)

File: plugins/utils/tmux-navigator.nix [hunk 5]
```diff
@@ -6,37 +6,126 @@
...
+            action = "<cmd><C-U>TmuxNavigateRight<CR>";
+            options.silent = true;
+          }
+          {
+            key = "<C-w>\\";
+            action = "<cmd><C-U>TmuxNavigatePrevious<CR>";
+            options.silent = true;
+          }
+        ];
+        ```
```
Comment (GaetanLepage) [inline plugins/utils/tmux-navigator.nix:126]:
```suggestion
        ```nix
          keymaps = [
            {
              key = "<C-w>h";
              action = "<cmd><C-U>TmuxNavigateLeft<CR>";
              options.silent = true;
            }
            {
              key = "<C-w>j";
              action = "<cmd><C-U>TmuxNavigateDown<CR>";
              options.silent = true;
            }
            {
              key = "<C-w>k";
              action = "<cmd><C-U>TmuxNavigateUp<CR>";
              options.silent = true;
            }
            {
              key = "<C-w>l";
              action = "<cmd><C-U>TmuxNavigateRight<CR>";
              options.silent = true;
            }
            {
              key = "<C-w>\\";
              action = "<cmd><C-U>TmuxNavigatePrevious<CR>";
              options.silent = true;
            }
          ];
        ```
```
Same, add an indent here.

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
Looks very good.
I have made a few comments.

File: plugins/utils/tmux-navigator.nix (hunk 1 above)
Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix:122]: Same for `<cr>`?

**Comment by GaetanLepage** [comment]
You will need to format your code by running `nix fmt`

File: plugins/utils/tmux-navigator.nix [hunk 6]
```diff
@@ -6,37 +6,127 @@
...
+    )
+    (
+      mkRemovedOptionModule
+      (option "tmuxNavigatorNoWrap")
+      "Use `${settingStr "no_wrap"}` option."
+    )
+  ];
+in
+  helpers.vim-plugin.mkVimPlugin config {
+    # FIXME document tmux-side installation instructions
```
Comment (MattSturgeon) [inline plugins/utils/tmux-navigator.nix:34]: I believe `mkVimPlugin` would need a "`extraEnableDesc`" parameter or similar, that way the default `"Whether to enable $name}."` description could be joined with extra documentation.

In this case something along the lines of `extraEnableDesc = "Note: tmux must be configured separately for tmux-navigator to work correctly. See [upstream docs](...) for instructions."`

I wonder if it's also worth having a small user-guide section covering this in more detail?

Comment (GaetanLepage) [reply plugins/utils/tmux-navigator.nix:34]: Good idea.
I made a PR: https://github.com/nix-community/nixvim/pull/1297

File: plugins/utils/tmux-navigator.nix [hunk 7]
```diff
@@ -6,37 +6,127 @@
...
+
+        This option disables those default mappings being created.
+
+        You can use the plugin's five commands to define your own custom mappings:
+
+        ```nix
+          keymaps = [
+            {
+              key = "<C-w>h";
+              action = "<cmd><C-U>TmuxNavigateLeft<cr>";
```
Comment (MattSturgeon) [inline plugins/utils/tmux-navigator.nix:103]: `<cmd><C-U>TmuxNavigateLeft<cr>` complains "not an editor command".

Looking at [`:h <cmd>`](https://neovim.io/doc/user/map.html#%3CCmd%3E), it seems like `<cmd>` alone does the same thing as `:<C-U>` while also working better with more modes.

The `<C-U>` is intended to clear the command prefix that may exist, e.g. when using `:` while a range is selected in visual mode. It's similar to using `<C-u>` in a terminal.

```suggestion
              action = "<cmd>TmuxNavigateLeft<cr>";
```

Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix:103]: Also:
> The command is not echo'ed, no need for `<silent>`.

File: plugins/utils/tmux-navigator.nix [hunk 8]
```diff
@@ -6,37 +6,127 @@
...
+      (option "tmuxNavigatorNoWrap")
+      "Use `${settingStr "no_wrap"}` option."
+    )
+  ];
+in
+  helpers.vim-plugin.mkVimPlugin config {
+    # FIXME document tmux-side installation instructions
+    name = "tmux-navigator";
+    originalName = "vim-tmux-navigator";
+    defaultPackage = pkgs.vimPlugins.tmux-navigator;
```
[the remark this answers is not in the document]
Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix:37]: I see, it's an [alias](https://github.com/NixOS/nixpkgs/blob/b06025f1533a1e07b6db3e75151caa155d1c7eb3/pkgs/applications/editors/vim/plugins/aliases.nix#L130). Do we want to continue using the alias or use the actual pkg name?

File: plugins/utils/tmux-navigator.nix [hunk 9]
```diff
@@ -6,37 +6,173 @@
...
+  ];
+in
+  helpers.vim-plugin.mkVimPlugin config {
+    name = "tmux-navigator";
+    originalName = "vim-tmux-navigator";
+    defaultPackage = pkgs.vimPlugins.vim-tmux-navigator;
+    globalPrefix = "tmux_navigator_";
 
-    package = helpers.mkPackageOption "tmux-navigator" pkgs.vimPlugins.tmux-navigator;
+    maintainers = [maintainers.GaetanLepage];
```
Comment (GaetanLepage) [inline plugins/utils/tmux-navigator.nix:39]: Wouldn't you want to take ownership of this plugin ?

Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix:39]: Happy to, it just felt a bit presumptuous. Can't promise active maintainership as time & interest fluctuate

Comment (GaetanLepage) [reply plugins/utils/tmux-navigator.nix:39]: Don't worry, this does not have a huge importance.

File: plugins/utils/tmux-navigator.nix (hunk 6 above)
Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix]: Rebased on main to make use of it :+1:

Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix]: Hm, getting newline errors...

```
$ nix build #docs --show-trace
error: builder for '/nix/store/2yc247fwdjc66324wgrrr421cvmnb4lc-nixvim-docs.drv' failed with exit code 2;
       last 2 log lines:
       > Running phase: buildPhase
       > /nix/store/3q4aagwsn8dp474143v2ixzqjrqibpv1-copy_docs: command substitution: line 3439: syntax error near unexpected token `newline'
```

Removing line breaks from the description allows docs to build... Are newlines supposed to work?

File: plugins/utils/tmux-navigator.nix [hunk 10]
```diff
@@ -6,37 +6,173 @@
...
-    package = helpers.mkPackageOption "tmux-navigator" pkgs.vimPlugins.tmux-navigator;
+    maintainers = [helpers.maintainers.MattSturgeon];
 
-    tmuxNavigatorSaveOnSwitch = helpers.mkNullOrOption (lib.types.enum [1 2]) ''
-      null: don't save on switch (default value)
-      1: update (write the current buffer, but only if changed)
-      2: wall (write all buffers)
-    '';
+    description = ''
+      When combined with a set of tmux key bindings, the plugin will allow you to navigate seamlessly between vim splits and tmux panes using a consistent set of hotkeys.
```
[the remark this answers is not in the document]
Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix:42]: Looks like an existing bug in the generated `copy_docs` script; the description (including title, url, maintainers) is `cat`ed along with the content of `options.md` into `index.md`, however the string is unescaped and <code>`</code> characters (etc) cause shell errors...

[the remark this answers is not in the document]
Comment (MattSturgeon) [reply plugins/utils/tmux-navigator.nix:42]: Fixed by 1889342524c707f16b4f44afe778627aa0f612cd

**Review by GaetanLepage — APPROVED** [review]
LGTM

File: docs/mdbook/default.nix [hunk 11]
```diff
@@ -196,6 +196,8 @@ with lib; let
             if isBranch
             then "${opts.index.path}/index.md"
             else "${opts.index.path}.md";
+          # Can't directly cat the moduleDoc text, so store it in a file
+          moduleDocMD = pkgs.writeText "${opts.index.path}/moduleDoc.md" opts.index.moduleDoc;
```
Comment (MattSturgeon) [inline docs/mdbook/default.nix:200]: If my understanding is correct, the filename passed to `writeText` doesn't need to be unique, so we can drop the `opts.index.path` prefix?

File: docs/mdbook/default.nix [hunk 12]
```diff
@@ -204,9 +206,7 @@ with lib; let
             then "cp ${mkMDDoc opts.index.options} ${path}"
             else ''
               {
-              cat <<EOF
-              ${opts.index.moduleDoc}
-              EOF
+              cat ${moduleDocMD}
```
Comment (GaetanLepage) [inline docs/mdbook/default.nix:209]: You can even inline it as so (the filename doesn't matter indeed).
```suggestion
              cat ${pkgs.writeText "module-doc" opts.index.moduleDoc}
```

File: docs/mdbook/default.nix (hunk 11 above)
Comment (MattSturgeon) [reply docs/mdbook/default.nix]: Pushed a fixup commit that can be squashed later

Comment (GaetanLepage) [reply docs/mdbook/default.nix]: You should keep the comment you had originally added.

File: docs/mdbook/default.nix (hunk 12 above)
Comment (MattSturgeon) [reply docs/mdbook/default.nix]: Done :+1:

File: docs/mdbook/default.nix [hunk 13]
```diff
@@ -204,9 +204,7 @@ with lib; let
             then "cp ${mkMDDoc opts.index.options} ${path}"
             else ''
               {
-              cat <<EOF
-              ${opts.index.moduleDoc}
-              EOF
+              cat ${pkgs.writeText "module-doc" opts.index.moduleDoc}
```
Comment (GaetanLepage) [inline docs/mdbook/default.nix:207]: Can you re-add the comment motivating this change please ?

Comment (MattSturgeon) [reply docs/mdbook/default.nix]: Sure thing

**Comment by GaetanLepage** [comment]
Ok for me.
I'll let @traxys proof-read as he'll have an external look.

**Review by traxys — APPROVED** [review]

**Pull request merged**